### PR TITLE
Add `tags` support to app

### PR DESCRIPTION
**Pull request by estrattonbailey**
## Notes
- currently using a fork of `@tiptap/suggestions`, awaiting https://github.com/ueberdosis/tiptap/pull/4515
- relying on local builds of unpublished backend PR https://github.com/bluesky-social/atproto/pull/1762

### CODE REVIEW

File: src/state/models/ui/tags-autocomplete.ts [hunk 1]
```diff
@@ -0,0 +1,115 @@
+import {makeAutoObservable, runInAction} from 'mobx'
+import AwaitLock from 'await-lock'
+import {RootStoreModel} from '../root-store'
+import Fuse from 'fuse.js'
+import {isObj, hasProp, isStrArray} from 'lib/type-guards'
+
+/**
+ * Used only to persist recent tags across app restarts.
+ */
```
Comment (estrattonbailey) [inline src/state/models/ui/tags-autocomplete.ts:9]: Not currently isolated per-account

File: src/state/models/ui/tags-autocomplete.ts [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+      // de-duplicates via Set
+      new Set([
+        // sample up to 3 recent tags
+        ...this.rootStore.recentTags.tags.slice(0, 3),
+        // sample up to 3 of your profile tags
+        ...this.profileTags.slice(0, 3),
+        // and all searched tags
+        ...this.searchedTags,
+      ]),
+    )
```
Comment (estrattonbailey) [inline src/state/models/ui/tags-autocomplete.ts:80]: The sampling for the tag type-ahead (profile tags are still todo)

File: src/view/com/composer/Composer.tsx [hunk 3]
```diff
@@ -228,6 +243,14 @@ export const ComposePost = observer(function ComposePost({
         imageCount: gallery.size,
       })
       if (replyTo && replyTo.uri) track('Post:Reply')
+
+      for (const facet of richtext.facets || []) {
+        for (const feature of facet.features) {
+          if (AppBskyRichtextFacet.isTag(feature)) {
+            tagsAutocompleteModel.commitRecentTag(feature.tag)
```
Comment (estrattonbailey) [inline src/view/com/composer/Composer.tsx:250]: Adds inline tags to recent tags model upon submission

File: src/view/com/composer/text-input/mobile/TagsAutocomplete.tsx [hunk 4]
```diff
@@ -0,0 +1,118 @@
+import React, {useEffect} from 'react'
+import {Animated, TouchableOpacity, StyleSheet, View} from 'react-native'
+import {observer} from 'mobx-react-lite'
+import {TagsAutocompleteModel} from 'state/models/ui/tags-autocomplete'
+import {useAnimatedValue} from 'lib/hooks/useAnimatedValue'
+import {usePalette} from 'lib/hooks/usePalette'
+import {Text} from 'view/com/util/text/Text'
+
+export function getHashtagAt(text: string, position: number) {
```
Comment (estrattonbailey) [inline src/view/com/composer/text-input/mobile/TagsAutocomplete.tsx:9]: This is roughly based on what Paul did for mentions

File: src/view/com/post-thread/PostThreadItem.tsx [hunk 5]
```diff
@@ -700,7 +723,7 @@ const useStyles = () => {
     },
     expandedInfo: {
       flexDirection: 'row',
-      padding: 10,
+      paddingVertical: 10,
```
Comment (estrattonbailey) [inline src/view/com/post-thread/PostThreadItem.tsx:726]: There's a lot of stuff in an expanded post item, and it looked better all left-aligned. However, we may want to adjust how we display all this bc it's kind of busy.

File: src/state/models/ui/tags-autocomplete.ts (hunk 1 above)
Comment (pfrazee) [reply src/state/models/ui/tags-autocomplete.ts:9]: Or persisted across devices

File: src/state/models/ui/tags-autocomplete.ts [hunk 6]
```diff
@@ -0,0 +1,128 @@
...
+import AwaitLock from 'await-lock'
+import {RootStoreModel} from '../root-store'
+import Fuse from 'fuse.js'
+import {isObj, hasProp, isStrArray} from 'lib/type-guards'
+
+/**
+ * Used only to persist recent tags across app restarts.
+ *
+ * TODO may want an LRU?
+ */
```
Comment (estrattonbailey) [inline src/state/models/ui/tags-autocomplete.ts:11]: Probably do, lmk if you disagree

File: src/view/com/composer/Composer.tsx [hunk 7]
```diff
@@ -211,6 +225,7 @@ export const ComposePost = observer(function ComposePost({
         onStateChange: setProcessingState,
         knownHandles: autocompleteView.knownHandles,
         langs: store.preferences.postLanguages,
+        tags,
```
Comment (estrattonbailey) [inline src/view/com/composer/Composer.tsx:228]: Outline tags are posted to API here, inline are contained in the raw text and detected later

File: src/view/com/composer/TagInput/index.tsx [hunk 8]
```diff
@@ -0,0 +1,268 @@
...
+    reset()
+    setIsInitialLoad(true)
+    input.current?.blur()
+  }, [reset, setIsInitialLoad])
+
+  return (
+    <View>
+      <TagInputEntryButton onRequestOpen={openSheet} tags={tags} />
+
+      <Portal>
```
Comment (estrattonbailey) [inline src/view/com/composer/TagInput/index.tsx:148]: New portal component

File: src/view/com/composer/text-input/TextInput.tsx [hunk 9]
```diff
@@ -94,17 +106,29 @@ export const TextInput = forwardRef(function TextInputImpl(
         newRt.detectFacetsWithoutResolution()
         setRichText(newRt)
 
-        const prefix = getMentionAt(
+        const mentionPrefix = getMentionAt(
```
Comment (estrattonbailey) [inline src/view/com/composer/text-input/TextInput.tsx:109]: Copied this for tags, so renamed for clarity

**Pull request closed without merging**